### PR TITLE
Bug 2020428: Adapt webpack 5 related code in dynamic plugin SDK

### LINKED ISSUE #9198: [dynamic-plugin-sdk] Plugin fails to load if "eval-source-map" webpack devtool is used
My plugin stopped working when I changed my webpack dev tool to `eval-source-map` (among a bunch of other things, so it was a lot of effort to debug :smile: )

It is because if you use the `eval-source-map` devtool (probably others too, this is just the one I happened to be using), webpack inserts a warning comment at the top of the file:

source-map-bug

but the dynamic-plugin-sdk assumes that `window.loadPluginEntry` is the first line of the file [here](https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts#L72).

so this results in the plugin ID being inserted at the wrong place (inside the comment, as highlighted in the screenshot above)

The error in the console is `Received callback for unknown plugin [object Object]`, 
followed by `Entry script for plugin <plugin id> loaded without callback`:
image

To compare, here it what it looks like when I switched to the `inline-source-map` devtool, you can see now the plugin ID is correctly passed as the first argument to `loadPluginEntry`.

source-map-working

### PR DESCRIPTION
**Pull request by vojtechszocs**
This PR addresses all TODO items and refactors plugin SDK webpack code to comply with webpack 5 APIs.

Closes #9198

### Improvements

- `ConsoleAssetPlugin` is now responsible for generating and/or post-processing all Console plugin assets.
- `ExtensionValidator` is now executed via [`emit`](https://webpack.js.org/api/compiler-hooks/#emit) hook. Plugins that wish to avoid emitting assets entirely due to webpack error(s) should do so via [`optimization.emitOnErrors`](https://webpack.js.org/configuration/optimization/#optimizationemitonerrors) option.
- `ExtensionValidator` can be disabled by passing `CONSOLE_PLUGIN_SKIP_EXT_VALIDATOR=true` env. variable to your webpack command, for example:
```
CONSOLE_PLUGIN_SKIP_EXT_VALIDATOR=true yarn ts-node node_modules/.bin/webpack
```
- Plugins should **_not_** specify [`output.publicPath`](https://webpack.js.org/configuration/output/#outputpublicpath) option. `ConsoleRemotePlugin` will always set its value to `/api/plugins/<ConsolePluginName>/`. A warning will be issued in case it's specified explicitly, for example:
```
<w> [ConsoleRemotePlugin] output.publicPath is defined, but will be overridden to /api/plugins/console-demo-plugin/
```
- exposed module chunks have better file names, e.g. `exposed-navPage-chunk.js`

### Shared module configuration

- all shared modules are treated as singletons, i.e. allow only a single version of the shared module
  - :arrow_right: Console is responsible for providing all shared modules to all of its plugins
- plugins won't bring their own fallback version of shared modules
  - :arrow_right: results in faster plugin builds and smaller plugin asset footprint
- no version checking when initializing shared modules for the given plugin
  - :arrow_right: using `*` semver range to ensure plugins always use the Console-provided shared module version

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/__tests__/shared-modules-init.spec.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    const [, entryModule] = getEntryModuleMocks({});
+
+    initSharedPluginModules(entryModule);
+
+    expect(entryModule.init).toHaveBeenCalledTimes(1);
+
+    expectSameValues(Object.keys(entryModule.init.mock.calls[0][0]), sharedPluginModules);
+  });
+
+  it('supports plugins built with an older version of plugin SDK', () => {
```
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/src/__tests__/shared-modules-init.spec.ts:20]: 👍

File: frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts [hunk 2]
```diff
@@ -48,63 +54,55 @@ export class ConsoleRemotePlugin {
...
-        library: { type: remoteEntryLibraryType, name: remoteEntryCallback },
-        filename: remoteEntryFile,
-        exposes: this.pkg.consolePlugin.exposedModules || {},
-      }).apply(compiler);
-
-      new webpack.sharing.SharePlugin({
-        shared: sharedPluginModules,
-      }).apply(compiler);
+    const logger = compiler.getInfrastructureLogger(ConsoleRemotePlugin.name);
+    const publicPath = `/api/plugins/${this.pkg.consolePlugin.name}/`;
```
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts:58]: Console has an option for a base context root, although we don't use it for OpenShift. Would this break that?

File: frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleAssetPlugin.ts [hunk 3]
```diff
@@ -29,62 +29,83 @@ export const validateExtensionsFileSchema = (
...
+          name: ConsoleAssetPlugin.name,
+          stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONS,
+        },
+        () => {
+          compilation.updateAsset(remoteEntryFile, (source) => {
+            const newSource = new webpack.sources.ReplaceSource(source);
+            newSource.insert(
+              this.remoteEntryCallback.length + 1,
+              `'${this.pkg.consolePlugin.name}@${this.pkg.consolePlugin.version}', `,
+            );
```
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleAssetPlugin.ts:87]: Not new, but see #9198.

**Review by spadgett — COMMENTED** [review]
We might need to revisit some of the assumptions we're making about shared modules down the road, but I think this is good for now.

File: frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts (hunk 2 above)
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts:58]: This option means that all plugin assets will be loaded from `<ConsoleOrigin>/api/plugins/<PluginName>` so relative to Console application's protocol / domain / port at runtime. Not sure if that is safe enough in relation to changing base context root :confused:

File: frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleAssetPlugin.ts (hunk 3 above)
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleAssetPlugin.ts:87]: OK, we can first find the position of `remoteEntryCallback` and insert from there.

**Comment by vojtechszocs** [comment]
Updated PR to also fix #9198

File: frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleAssetPlugin.ts (hunk 3 above)
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleAssetPlugin.ts]: PR updated to address this issue.

File: frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts (hunk 2 above)
Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts:58]: No, that will break context roots. We might need it for HAC as well :/

Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts:58]: https://github.com/openshift/console/blob/master/cmd/bridge/main.go#L69

**Review by spadgett — APPROVED** [review]
/lgtm

**Pull request merged**